python/lsst/daf/butler/sql/multipleDatasetQueryBuilder.py

```diff
@@ -0,0 +1,373 @@
+            A subquery for this `DatasetType` must have already been added to
+            the query via `fromDatasetTypes` or `joinDatasetType`.
+        """
+        self.resultColumns.addDatasetId(self._subqueries[datasetType].subquery, datasetType)
+
+    def findSelectableForLink(self, link):
+        # Docstring inherited from QueryBuilder.findSelectableForLink
+        result = super().findSelectableForLink(link)
+        if result is None:
+            for datasetType, data in self._subqueries.items():
+                if data.necessity is DatasetNecessityEnum.REQUIRED and link in data.links:
+                    result = data.subquery
+                    break
+        return result
+
+    def findSelectableByName(self, name):
+        # Docstring inherited from QueryBuilder.findSelectableByName
+        result = super().findSelectableByName(name)
+        if result is None:
+            for datasetType, data in self._subqueries.items():
+                if name == datasetType.name:
+                    result = data.subquery
+                    break
+        return result
+
+    def convertResultRow(self, managed, *, expandDataIds=True):
+        r"""Convert a result row for this query to a `MultipleDatasetQueryRow`.
+
+        Parameters
+        ----------
+        managed : `ResultColumnsManager.ManagedRow`
+            Intermediate result row object to convert.
+        expandDataIds : `True`
+            If `True` (default), query the registry again to fully populate
+            all `DataId`s associated with `DatasetRef`s.  The full-row data
+            ID is never expanded.
+
+        Returns
+        -------
+        row : `MultipleDatasetQueryRow`
+            Object containing the `DataId`s and `DatasetRefs` produced by the
+            query.
+        """
+        dataId = managed.makeDataId()
+        datasetRefs = {}
+        for datasetType, data in self._subqueries.items():
+            ref = managed.makeDatasetRef(datasetType, expandDataId=expandDataIds)
+            if ref.id is None and data.necessity is DatasetNecessityEnum.PREREQUISITE:
+                raise LookupError(f"Search failed for prerequisite dataset "
+                                  f"{datasetType.name} associated with data ID {dataId}.")
+            datasetRefs[datasetType] = ref
+        for datasetType, data in self._deferrals.items():
+            # Do the deferred dataset query, by running essentially the same
+            # SQL we would have used as a subquery before, but with the
+            # JOIN..ON expression now used as a WHERE expression with explicit
+            # values from the row's data ID.
+            expr = []
+            for link in data.joinLinks:
+                selectable = data.builder.findSelectableForLink(link)
+                column = selectable.c[link]
+                expr.append(column == dataId[link])
+            ref = data.builder.executeOne(whereSql=and_(*expr))
```

Comment:
The other code path that does include these queries in the big one instead of deferring them is still alive, and should Just Work for databases other than SQLite (where we need to defer to avoid the 64-term join limit), and I'm hoping we can avoid scaling problems with SQLite by just not using SQLite at scale.

Right now we also always do other follow-up queries to obtain component datasets for composites, of course, but that's a separate issue.